File: CppLox/Core/Scanner.cpp
#include "Scanner.hpp"
#include "ScannerError.h"

std::vector<Token> Scanner::scan_tokens()
{
    while(!isAtEnd())
    {
        start_ = current_;
        scanToken();
    }
    string str;
    tokens_.emplace_back(TokenType::EOF, str, str, line_);
    return tokens_;
}

std::vector<IError*> Scanner::get_errors()
{
    return errors_;
}

bool Scanner::isAtEnd()
{
    return current_ >= source_.size();
}

void Scanner::addToken(TokenType type)
{
    string str;
    addToken(type, str);
}

void Scanner::addToken(TokenType type, string& literal)
{
    string text = source_.substr(start_, current_ - start_);
    tokens_.emplace_back(type, text, literal, line_);
}

void Scanner::addError(string text)
{
    //errors_.push_back(std::make_unique<ScannerError>(line_, text));
}

char Scanner::peek()
{
    return isAtEnd() ? '\0' : source_.at(current_);
}

void Scanner::add_string()
{
    while(peek() != '"' && !isAtEnd())
    {
        if(peek() == '\n')
            ++line_;
        advance();
    }

    // Unterminated string.
    if(isAtEnd())
    {
        addError("Unterminated string.");
        return;
    }

    //The closing ".
    advance();

    // Trim the surrounding quotes
    std::string value = source_.substr(start_ + 1, current_ - start_ - 2);
    addToken(TokenType::STRING, value);
}

char Scanner::peekNext()
{
    if(current_ + 1 >= source_.length())
        return '\0';
    return source_.at(current_ + 1);
}

void Scanner::add_number()
{
    while(isDigit(peek()))
        advance();

    // Look for a fractional part.
    if(peek() == '.' && isDigit(peekNext()))
    {
        // Consume the "."
        advance();
        while(isDigit(peek()))
            advance();
    }
    string literal = source_.substr(start_, current_);
    addToken(TokenType::NUMBER, literal);
}

bool Scanner::isDigit(char c)
{
    return c >= '0' && c <= '9';
}

bool Scanner::isAlphaNumeric(char c)
{
    return isAlpha(c) || isDigit(c);
}

void Scanner::add_identifier()
{
    while(isAlphaNumeric(peek()))
        advance();

    string text = source_.substr(start_, current_ - start_);
    const auto it = keywords_.find(text);
    TokenType type = TokenType::IDENTIFIER;
    if(it != keywords_.end())
        type = it->second;

    addToken(type);
}

bool Scanner::isAlpha(char c)
{
    return c >= 'a' && c <= 'z' ||
           c >= 'A' && c <= 'Z' ||
           c == '_';
}

void Scanner::scanToken()
{
    char c = advance();
    switch(c)
    {
    case '(':
        addToken(TokenType::LEFT_PAREN);
        break;
    case ')':
        addToken(TokenType::RIGHT_PAREN);
        break;
    case '{':
        addToken(TokenType::LEFT_BRACE);
        break;
    case '}':
        addToken(TokenType::RIGHT_BRACE);
        break;
    case ',':
        addToken(TokenType::COMMA);
        break;
    case '.':
        addToken(TokenType::DOT);
        break;
    case '-':
        addToken(TokenType::MINUS);
        break;
    case '+':
        addToken(TokenType::PLUS);
        break;
    case ';':
        addToken(TokenType::SEMICOLON);
        break;
    case '*':
        addToken(TokenType::STAR);
        break;
    case '!':
        addToken(match('=') ? TokenType::BANG_EQUAL : TokenType::BANG);
        break;
    case '=':
        addToken(match('=') ? TokenType::EQUAL_EQUAL : TokenType::EQUAL);
        break;
    case '<':
        addToken(match('=') ? TokenType::LESS_EQUAL : TokenType::LESS);
        break;
    case '>':
        addToken(match('=') ? TokenType::GREATER_EQUAL : TokenType::EQUAL);
        break;
    case 'o':
        if(peek() == 'r')
        {
            addToken(TokenType::OR);
        }
        break;
    case '/':
    {
        if(match('/'))
        {
            while(peek() != '\n' && !isAtEnd())
                advance();
        }
        else if(isAlpha(c))
        {
            add_identifier();
        }
        else
        {
            addToken(TokenType::SLASH);
        }
        break;
    }
    case '"':
        add_string();
        break;
    case ' ':
    case '\r':
    case '\t':
        // Ignore whitespace.
        break;
    case '\n':
        ++line_;
        break;
    default:
        if(isDigit(c))
        {
            add_number();
        }
        else if(isAlpha(c))
        {
            add_identifier();
        }
        else
        {
            addError("Unexpected character.");
            break;
        }
    }
}

char Scanner::advance()
{
    ++current_;
    return source_.at(current_ - 1);
}

bool Scanner::match(char expected)
{
    if(isAtEnd())
        return false;
    if(source_.at(current_) != expected)
        return false;

    ++current_;
    return true;
}

File: Cslox/Scanner.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace Cslox
{
    internal sealed class Scanner
    {
        private readonly string _source;
        private readonly List<Token> _tokens = new List<Token>();
        private readonly Dictionary<string, TokenType> _keywords = new Dictionary<string, TokenType>();
        private readonly List<string> _errors;
        private int _start, _current, _line;

        public Scanner(in string source)
        {
            _source = source;
            _errors = new List<string>();
            InitKeywords();
        }

        public IReadOnlyList<string> GetErrors()
        {
            return _errors;
        }

        private void InitKeywords()
        {
            _keywords.Add("and", TokenType.And);
            _keywords.Add("class", TokenType.Class);
            _keywords.Add("else", TokenType.Else);
            _keywords.Add("false", TokenType.False);
            _keywords.Add("for", TokenType.For);
            _keywords.Add("fun", TokenType.Fun);
            _keywords.Add("if", TokenType.If);
            _keywords.Add("nil", TokenType.Nil);
            _keywords.Add("or", TokenType.Or);
            _keywords.Add("print", TokenType.Print);
            _keywords.Add("return", TokenType.Return);
            _keywords.Add("super", TokenType.Super);
            _keywords.Add("this", TokenType.This);
            _keywords.Add("true", TokenType.True);
            _keywords.Add("var", TokenType.Var);
            _keywords.Add("while", TokenType.While);
        }

        public IReadOnlyList<Token> ScanTokens()
        {
            while (!IsAtEnd())
            {
                try
                {
                    _start = _current;
                    ScanToken();
                }
                catch (Exception e)
                {
                    _errors.Add($"[Scanning Error] {e.Message} at line: {_line + 1}\n {_source.Split('\n')[_line]}");
                }
            }

            _tokens.Add(new Token(TokenType.Eof, "", null));
            return _tokens;
        }

        private void ScanToken()
        {
            var ch = Advance();

            switch (ch)
            {
                case '(':
                    AddToken(TokenType.LeftParen);
                    break;
                case ')':
                    AddToken(TokenType.RightParen);
                    break;
                case '{':
                    AddToken(TokenType.LeftBrace);
                    break;
                case '}':
                    AddToken(TokenType.RightBrace);
                    break;
                case ',':
                    AddToken(TokenType.Comma);
                    break;
                case '.':
                    AddToken(TokenType.Dot);
                    break;
                case '-':
                    AddToken(TokenType.Minus);
                    break;
                case '+':
                    AddToken(TokenType.Plus);
                    break;
                case ';':
                    AddToken(TokenType.Semicolon);
                    break;
                case '*':
                    AddToken(TokenType.Star);
                    break;
                case '!':
                    AddToken(Match('=') ? TokenType.BangEqual : TokenType.Bang);
                    break;
                case '=':
                    AddToken(Match('=') ? TokenType.EqualEqual : TokenType.Equal);
                    break;
                case '<':
                    AddToken(Match('=') ? TokenType.LessEqual : TokenType.Less);
                    break;
                case '>':
                    AddToken(Match('=') ? TokenType.GreaterEqual : TokenType.Greater);
                    break;
                case '/':
                {
                    if (Match('/'))
                    {
                        while (Peek() != '\n' && !IsAtEnd())
                            Advance();
                        break;
                    }

                    if (Match('*'))
                    {
                        do
                        {
                            Advance();
                            if (Peek() == '\n')
                            {
                                ++_line;
                                continue;
                            }

                            if (Peek() == '*')
                            {
                                Advance();
                                if (Peek() == '/')
                                {
                                    Advance();
                                    return;
                                }
                            }
                        } while (!IsAtEnd());

                        throw new Exception("Multiline comments must be closed");
                    }

                    AddToken(TokenType.Slash);
                    break;
                }
                case '"':
                    CreateString();
                    break;

                case ' ':
                case '\r':
                case '\t':
                    return;
                case '\n':
                    _line++;
                    return;
                default:
                    if (IsDigit(ch))
                    {
                        CreateNumber();
                    }
                    else if (IsAlpha(ch))
                    {
                        CreateIdentifier();
                    }
                    else
                    {
                        throw new Exception("Unexpected symbol");
                    }

                    break;
            }
        }

        private void CreateIdentifier()
        {
            while (IsAlphaNumeric(Peek()))
                Advance();

            AddToken(_keywords.TryGetValue(CurrentLexem(), out var type) ? type : TokenType.Identifier);
        }

        private static bool IsAlphaNumeric(char ch)
        {
            return IsAlpha(ch) || IsDigit(ch);
        }

        private static bool IsAlpha(char ch)
        {
            return ch >= 'a' && ch <= 'z' ||
                   ch >= 'A' && ch <= 'Z' ||
                   ch == '_';
        }

        private void CreateNumber()
        {
            while (IsDigit(Peek()))
                Advance();
            if (Peek() == '.' && IsDigit(PeekNext()))
            {
                Advance();
                while (IsDigit(Peek())) Advance();
            }


            if (double.TryParse(CurrentLexem(), NumberStyles.Number, CultureInfo.InvariantCulture, out var literal))
                AddToken(TokenType.Number, literal);
            else
                throw new Exception("Unexpected symbol");
        }

        private string CurrentLexem()
        {
            return _source.Substring(_start, _current - _start);
        }

        private char PeekNext()
        {
            return _current + 1 >= _source.Length ? '\0' : _source[_current + 1];
        }

        private static bool IsDigit(char ch)
        {
            return ch >= '0' && ch <= '9';
        }

        private void CreateString()
        {
            while (Peek() != '"' && !IsAtEnd())
            {
                if (Peek() == '\n') _line++;
                Advance();
            }

            if (IsAtEnd())
            {
                throw new Exception("Unterminated string");
            }

            Advance();
            var value = _source.Substring(_start, _current - _start);
            AddToken(TokenType.String, value);
        }

        private char Peek()
        {
            return IsAtEnd() ? '\0' : _source[_current];
        }

        private bool Match(char expected)
        {
            if (IsAtEnd() || _source[_current] != expected)
                return false;
            _current++;
            return true;
        }

        private void AddToken(TokenType type, object literal = null)
        {
            var text = _source.Substring(_start, _current - _start);
            _tokens.Add(new Token(type, text, literal));
        }

        private char Advance()
        {
            _current++;
            return _source[_current - 1];
        }

        private bool IsAtEnd()
        {
            return _current >= _source.Length;
        }
    }
}